### DISCUSSION TITLE
ATTiny424 DDRA/DDRB support?

### DISCUSSION THREAD

**Discussion by bathtub-toast** [SpenceKonde/megaTinyCore#687] [Q&A] [answered]
Using just the arduino and trying to build my code it tells me these are undefined.
I'm willing to put in the time to try and create it, if anyone can point me a file that actually works, since I can't find where the DDRB is defined for the Arduino Uno or Leonardos source code.
The pin groups exist, per atmels whitepaper, but its just not clear to me how I might actually implement them.
The PORTA and PORTB appear to be implemented and working correctly, which is why I'm only looking for DDRA and DDRB

Anyone with ideas or working examples would be awesome.

**Comment by Modac** [comment, marked as answer]
Have you read [this](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_DirectPortManipulation.md) yet?
Especially [this](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_DirectPortManipulation.md#equivalents-to-classic-avr-registers) table.

**Reply by bathtub-toast** [reply]
I must admit I completely missed this page.
Thank you!

**Reply by bathtub-toast** [reply]
Just to make sure I'm not confused I can still pass my same byte array values that DDRx would've used? I have something like 0b00000110 for turning on PORTA 1/2 as output and setting the rest as input.

The charlieplex 'example' I found explained nothing on how any of this worked, and I have to admit I didn't know it was actually called Direct Port Manipulation, which makes far more sense that what I thought it was doing.

**Reply by bathtub-toast** [reply]
Just verifying that I didn't misread this line
`PORTA.DIR = 0b01001000;         // sets PA3 and PA6 as an output`

**Reply by Modac** [reply]
Yeah that should be right, if I'm not mistaken the only major thing not interchangable between the old and new registers is setting the input pullups.

A small remark, I like these lines better for setting the pin direction:
`PORTA.DIR = PIN3_bm | PIN6_bm;  // sets PA3 and PA6 as an output`
or
`PORTA.DIRSET = PIN3_bm | PIN6_bm;  // sets PA3 and PA6 as an output, does not change other pins`
I just think they are more readable and I think the second one also has some advantage but I'm not too sure. 😄

**Comment by SpenceKonde** [comment]
no, that's correct (pins are zero indexed, and the rightmotst bit is bit 0.

**Reply by SpenceKonde** [reply]
See the port section of the IO headers for the full list of names The only way to program the modern parts is with a text editor, the io header open so you know how everything is spelled a
https://github.com/SpenceKonde/megaTinyCore/tree/master/megaavr/extras/ioheaders
Functions are covered by documents already linked above. 

And of course the 400 page datasheet you routinely wish were more detailed. 

The new port system is actually a massive improvement from classic AVR, because we get two bonuses:
PORTx.DIRSET/DIRCLR/DIRTGL registers (and corresponding OUT* registers

```c++
PORTA.DIR = 0b01001001;  //2 clock cycles plus 1 to LDI  if not in memory
// now I want to set bit 4. I can:
// assume it hasn't beet changed because I wrote the code and know nothing else changes it.
PORTA.DIR = 0b01011001;  //2 clock cycles plus 1 to LDI  if not in memory
// but that's not something that you can normally count on
PORTA.DIR = PORTA.DIR | 0b00010000;  5(typically) 6 clock cycles, and is NON-ATOMIC. Making it atomic takes 3 more clocks, 2 if you're 1000% sure it will never be called from an interrupt
// 6 cycles!? That sucks!
// They give faster alternatives. 
PORTA.DIRSET = 0b0001000; // 2 clockls
PORTA.DIRCLR = 0b0001000; // 2 clock
PORTA.DIRTGL = 0b0001000; / /2 clocks 
//2 clock cycles typical. must only be used with =, not |= or &=, otherwise you will attract the attention of demons*

 Atomic write 
VPORTA.DIR = 0b01011001// 1 clock cycle plus the LDI
//But if you want to get the fastest writes and know pin at compile time (ideally whether it's being written HIGH OR LOW) 
VPORTA.DIR |= (1<<0x03); // single cycle, nothing LDI'ed to write out. Atomic.

//But thou shalt never use VPORTs with multiple bits and |= or &=:
VPORTA.DIR |= (3<<0x3); // 3 clocks non atomic!!!

`*` You know how they say the devil is in the details. in this case, he's located in the fact that all registers READ the same value, so the compiler 

The cost is that the pullup control bit has been moved to PORTx.PINnCTRL - but that gives us other neat possibilities, See https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_Digital.md
